mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp

```diff
@@ -91,6 +117,330 @@ void mlir::configureGpuToROCDLConversionLegality(ConversionTarget &target) {
+        loc, LLVMI32Type, ArrayRef<Value>{voffsetElements, elementBytes});
+
+    // populate slc : fixed as 0 of type i1.
+    Value slc = rewriter.create<LLVM::ConstantOp>(
+        loc, LLVMI1Type, rewriter.getIntegerAttr(I1Type, 0));
+
+    // populate glc : fixed as 0 of type i1.
+    Value glc = rewriter.create<LLVM::ConstantOp>(
+        loc, LLVMI1Type, rewriter.getIntegerAttr(I1Type, 0));
+
+    if (srcElementType.getIntOrFloatBitWidth() != 32) {
+      // for f16 and i16 (bf16) types, use f32 buffer_load and bitcast the
+      // result.
+      assert(resultType.isa<VectorType>());
+      // deduce the interim type for f16 / i16 (bf16).
+      VectorType vectorResultType = resultType.template cast<VectorType>();
+      auto vectorShape = vectorResultType.getShape();
+
+      SmallVector<int64_t, 1> interimShape;
+      for (unsigned iter = 0; iter < vectorShape.size() - 1; ++iter)
+        interimShape.push_back(vectorShape[iter]);
+      interimShape.push_back(vectorShape[vectorShape.size() - 1] >> 1);
+      bool useScalarF32 = (vectorShape.size() == 1) && (vectorShape[0] == 2);
+      Type interimResultType;
+      if (useScalarF32)
+        interimResultType = rewriter.getF32Type();
+      else
+        interimResultType =
+            VectorType::get(interimShape, rewriter.getF32Type());
+      Type interimLLVMResultType =
+          typeConverter->convertType(interimResultType);
+
+      Value interimLoad = rewriter.create<ROCDL::MubufLoadOp>(
+          loc, interimLLVMResultType, rsrc, vindex, voffset, slc, glc);
+
+      rewriter.replaceOpWithNewOp<LLVM::BitcastOp>(op, LLVMResultType,
+                                                   interimLoad);
+    } else {
+      rewriter.replaceOpWithNewOp<ROCDL::MubufLoadOp>(
+          op, LLVMResultType, rsrc, vindex, voffset, slc, glc);
+    }
+
+    return success();
+  }
+};
+
+struct MubufStoreOpLowering : ConvertToLLVMPattern {
+  explicit MubufStoreOpLowering(MLIRContext *context,
+                                LLVMTypeConverter &typeConverter)
+      : ConvertToLLVMPattern(gpu::MubufStoreOp::getOperationName(), context,
+                             typeConverter) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto mubufStoreOp = cast<gpu::MubufStoreOp>(op);
+    auto adaptor = gpu::MubufStoreOpAdaptor(operands);
+    auto loc = mubufStoreOp.getLoc();
+
+    MemRefType dstMemRefType =
+        mubufStoreOp.memref().getType().cast<MemRefType>();
+    Type dstElementType = dstMemRefType.getElementType();
+    auto dstShape = dstMemRefType.getShape();
+    auto adaptorIndices = adaptor.indices();
+    auto adaptorValue = adaptor.value();
+
+    Type valueType = mubufStoreOp.value().getType();
+    Type LLVMValueType = typeConverter->convertType(valueType);
+
+    // use standard store for storing scalar f16 and i16 (bf16).
```

Comment:
Do we need to pack into a reg for < 32-bit values in order to use mubuf.store?

Reply:
the eventual ISA we get is `buffer_store_dword`, so the smallest unit is a DWORD, a 32-bit value. here we use `std.store` in MLIR std dialect, which would give us `flat_store` or `global_store` instrucitons which supports 16-bit values natively.